Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorThresholdEffect.py

```diff
@@ -444,6 +633,206 @@ def preview(self):
+      self.histogramPipeline.addPoint(ras)
+      self.updateHistogram()
+      abortEvent = True
+    elif eventId == vtk.vtkCommand.LeftButtonReleaseEvent:
+      self.histogramPipeline.state = HISTOGRAM_STATE_PLACED
+      abortEvent = True
+    elif eventId == vtk.vtkCommand.MouseMoveEvent:
+      if self.histogramPipeline.state == HISTOGRAM_STATE_MOVING:
+        self.histogramPipeline.addPoint(ras)
+        self.updateHistogram()
+        abortEvent = True
+    else:
+      pass
+    return abortEvent
+
+  def createHistogramPipeline(self, sliceWidget):
+    brushType = HISTOGRAM_TYPE_CIRCLE
+    if self.boxROIButton.checked:
+      brushType = HISTOGRAM_TYPE_BOX
+    elif self.drawROIButton.checked:
+      brushType = HISTOGRAM_TYPE_DRAW
+    elif self.lineROIButton.checked:
+      brushType = HISTOGRAM_TYPE_LINE
+    pipeline = HistogramPipeline(self, self.scriptedEffect, sliceWidget, brushType)
+    self.histogramPipeline = pipeline
+
+  def processViewNodeEvents(self, callerViewNode, eventId, viewWidget):
+    if self.histogramPipeline is not None:
+      self.histogramPipeline.updateBrushModel()
+
+  def onHistogramMouseClick(self, pos, button):
+    self.selectionStartPosition = pos
+    self.selectionEndPosition = pos
+    if (button == qt.Qt.RightButton):
+      self.selectionStartPosition = None
+      self.selectionEndPosition = None
+      self.minMaxFunction.RemoveAllPoints()
+      self.averageFunction.RemoveAllPoints()
+    self.updateHistogram()
+
+  def onHistogramMouseMove(self, pos, button):
+    self.selectionEndPosition = pos
+    if (button == qt.Qt.RightButton):
+      return
+    self.updateHistogram()
+
+  def onHistogramMouseRelease(self, pos, button):
+    self.selectionEndPosition = pos
+    if (button == qt.Qt.RightButton):
+      return
+    self.updateHistogram()
+
+  def updateHistogram(self):
+    masterImageData = self.scriptedEffect.masterVolumeImageData()
+    if masterImageData is None or self.histogramPipeline is None:
+      self.histogramFunction.RemoveAllPoints()
+      return
+
+    # Ensure that the brush is in the correct location
+    self.histogramPipeline.updateBrushModel()
+
+    self.stencil.SetInputConnection(self.histogramPipeline.worldToSliceTransformer.GetOutputPort())
+
+    self.histogramPipeline.worldToSliceTransformer.Update()
+    brushPolydata = self.histogramPipeline.worldToSliceTransformer.GetOutput()
+    brushBounds = brushPolydata.GetBounds()
+    brushExtent = [0, -1, 0, -1, 0, -1]
+    for i in range(3):
+      brushExtent[2*i] = vtk.vtkMath.Floor(brushBounds[2*i])
+      brushExtent[2*i+1] = vtk.vtkMath.Ceil(brushBounds[2*i+1])
+    if brushExtent[0] > brushExtent[1] or brushExtent[2] > brushExtent[3] or brushExtent[4] > brushExtent[5]:
+      self.histogramFunction.RemoveAllPoints()
+      return
+
+    sliceLogic = self.histogramPipeline.sliceWidget.sliceLogic()
+    backgroundLogic = sliceLogic.GetBackgroundLayer()
```

Comment:
Usually the master volume is the background volume but the user can change that. Maybe we could choose between foreground/background depending on which one is the master volume and if neither then log a warning and use the more visible one.

Reply:
Sure. Currently the background is used for the threshold preview. Should that be changed as well?

Reply:
Yes, it would be nice to use the same logic for both.